File: quadrotor/gumstix/src/command_data.c
/*
 * Copyright (c) Harald Roeck hroeck@cs.uni-salzburg.at
 * Copyright (c) Rainer Trummer rtrummer@cs.uni-salzburg.at
 *
 * University Salzburg, www.uni-salzburg.at
 * Department of Computer Science, cs.uni-salzburg.at
 */

/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include <stdio.h>

#include "command_data.h"


int command_data_to_stream( const command_data_t *data, char *buf, int len )
{
    if( len == COMMAND_DATA_SIZE )
    {
        buf[0] = (char)( data->roll >> 8 );
        buf[1] = (char)( data->roll );
        buf[2] = (char)( data->pitch >> 8 );
        buf[3] = (char)( data->pitch );
        buf[4] = (char)( data->yaw >> 8 );
        buf[5] = (char)( data->yaw );
        buf[6] = (char)( data->z >> 8 );
        buf[7] = (char)( data->z );
        return( 0 );
    }

    fprintf( stderr, "ERROR in %s %d: invalid length (%d) of command data to stream\n",
        __FILE__, __LINE__, len );
    return( -1 );
}

int command_data_from_stream( command_data_t *data, const char *buf, int len )
{
    if( len == COMMAND_DATA_SIZE )
    {
        data->roll  = (int16_t)( (buf[0] << 8) | (buf[1] & 0xFF) );
        data->pitch = (int16_t)( (buf[2] << 8) | (buf[3] & 0xFF) );
        data->yaw   = (int16_t)( (buf[4] << 8) | (buf[5] & 0xFF) );
        data->z     = (int16_t)( (buf[6] << 8) | (buf[7] & 0xFF) );
        return( 0 );
    }

    fprintf( stderr, "ERROR in %s %d: invalid length (%d) of command data from stream\n",
        __FILE__, __LINE__, len );
    return( -1 );
}

/* End of file */

File: quadrotor/ControlTerminal/src/javiator/util/CommandData.java
/*****************************************************************************/
/*   This code is part of the JAviator project: javiator.cs.uni-salzburg.at  */
/*                                                                           */
/*   Copyright (c) 2006  Harald Roeck, Rainer Trummer                        */
/*                                                                           */
/*   This program is free software; you can redistribute it and/or modify    */
/*   it under the terms of the GNU General Public License as published by    */
/*   the Free Software Foundation; either version 2 of the License, or       */
/*   (at your option) any later version.                                     */
/*                                                                           */
/*   This program is distributed in the hope that it will be useful,         */
/*   but WITHOUT ANY WARRANTY; without even the implied warranty of          */
/*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           */
/*   GNU General Public License for more details.                            */
/*                                                                           */
/*   You should have received a copy of the GNU General Public License       */
/*   along with this program; if not, write to the Free Software Foundation, */
/*   Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.      */
/*                                                                           */
/*****************************************************************************/

package javiator.util;

/**
 * Represents the navigation data
 * @author hroeck 
 */
public class CommandData extends NumeratedSendable
{
	public static final int PACKET_SIZE = 8;
	
	public CommandData( )
    {
        reset();
    }

    public void reset( )
    {
        roll  = 0;
        pitch = 0;
        yaw   = 0;
        z     = 0;
    }

    public String toString( )
    {
        String result;
        result  = " roll: "  + roll;
        result += " pitch: " + pitch;
        result += " yaw: "   + yaw;
        result += " z: "     + z;

        return( result );
    }

    public CommandData( Packet packet )
    {
        fromPacket( packet );
    }

    public synchronized Packet toPacket( )
    {
        Packet packet = new Packet( PACKET_SIZE );
        encode( packet, 0 );

        return( packet );
    }

    public void encode( Packet packet, int offset )
	{
		packet.payload[ offset + 0 ] = (byte)( roll >> 8 );
		packet.payload[ offset + 1 ] = (byte)( roll );
		packet.payload[ offset + 2 ] = (byte)( pitch >> 8 );
		packet.payload[ offset + 3 ] = (byte)( pitch );
		packet.payload[ offset + 4 ] = (byte)( yaw >> 8 );
		packet.payload[ offset + 5 ] = (byte)( yaw );
		packet.payload[ offset + 6 ] = (byte)( z >> 8 );
		packet.payload[ offset + 7 ] = (byte)( z );
	}

    public synchronized Packet toPacket( byte type )
    {
        Packet packet = toPacket( );
        packet.type   = type;
        packet.calcChecksum( );

        return( packet );
    }

    public synchronized void fromPacket( Packet packet )
    {
        decode( packet, 0 );
    }

	public void decode( Packet packet, int offset )
	{
		roll  = (short)( (packet.payload[ offset + 0 ] << 8) | (packet.payload[ offset + 1 ] & 0xFF) );
		pitch = (short)( (packet.payload[ offset + 2 ] << 8) | (packet.payload[ offset + 3 ] & 0xFF) );
		yaw   = (short)( (packet.payload[ offset + 4 ] << 8) | (packet.payload[ offset + 5 ] & 0xFF) );
		z     = (short)( (packet.payload[ offset + 6 ] << 8) | (packet.payload[ offset + 7 ] & 0xFF) );
	}

    public synchronized Object clone( )
    {
        CommandData copy = null;
        copy = (CommandData) super.clone( );

        if( copy != null )
        {
        	copy.roll  = roll;
        	copy.pitch = pitch;
        	copy.yaw   = yaw;
        	copy.z     = z;
        }

        return( copy );
    }

    public CommandData deepClone( )
    {
        return (CommandData) clone( );
    }

    public void copyTo( Copyable to )
	{
    	super.copyTo( to );
    	CommandData copy = (CommandData) to;
    	copy.roll  = roll;
    	copy.pitch = pitch;
    	copy.yaw   = yaw;
    	copy.z     = z;
	}
    
    public short roll;
    public short pitch;
    public short yaw;
    public short z;
}
